[ORM] Refactor expresssion builder to add joins when needed

GSadee opened the pull request:
I'm not sure if it is a good base branch for this PR.

This change is necessary for filtering by a nested entity: https://github.com/Sylius/Sylius/pull/10382/files#diff-decadb1c0264a2aa9b1d8a2b521fbbf2R49

Zales0123 commented:
I see three ways:

- we could make a detailed PHPSpec test for that (but it can be a pain in the neck, to be honest)
- we could create a unit test with PHPUnit with some most well-known use-cases (one field without nesting, two fields with one-level nesting, one field with two-levels nesting, etc.)
- indeed create some test application, but we must be aware that it would require a lot of additional code - it's not a plugin but a bundle, it does not have a test application already inside

I would personally give a try for the 2nd way (even though I'm not 100% sure will it work smoothly 🚀) and maybe think about other ways it fails. cc @pamil your opinion would be significant 🖖

lchrusciel commented:
Just for the record, there is a sample app: https://github.com/Sylius/SyliusGridBundle/tree/master/src/Bundle/test

Zales0123 commented:
My bad 😄 🖖 Still, the PHPUnit test can be sufficient (or not, but we should check it out)

GSadee commented:
For now, I've added a PHPSpec test and I hope it can be merged in this form.

Additionally, I've added an issue with missing proper tests (https://github.com/Sylius/SyliusGridBundle/issues/4) and I'll try to take care of it in the nearest future.

pamil commented on src/Bundle/Doctrine/ORM/ExpressionBuilder.php:
```diff
@@ -27,188 +27,135 @@ public function __construct(QueryBuilder $queryBuilder)
-        }
-
+        $field = $this->adjustField($field);
         $parameterName = $this->getParameterName($field);
```
Are we sure that we want to run `$this->getParameterName()` with **adjusted field** as an argument? Isn't it changing the previous behaviour and breaking BC?

GSadee replied:
`adjustField` is a method that coheres the given field, because field could look like `o.order` or `order`. So yes, the name returned by `getParameterName` could be now different (e.q. `order` instead of `o_order`), because I always remove the root alias, but I don't think that this is a BC break. The same grid configuration will work in the same way as it was before.

lchrusciel commented on src/Component/Provider/ArrayGridProvider.php:
```diff
@@ -57,6 +57,6 @@ private function extend(array $gridConfiguration, array $parentGridConfiguration
         unset($configuration['extends']);
 
-        return $configuration;
+        return $configuration !== null ? $configuration : [];
```
Won't it be the same?
```suggestion
        return $configuration ?? [];
```

lchrusciel commented on src/Bundle/Doctrine/ORM/ExpressionBuilder.php:
```diff
@@ -27,164 +27,135 @@ public function __construct(QueryBuilder $queryBuilder)
     {
         if (false === strpos($field, '.')) {
-            return $this->queryBuilder->getRootAlias() . '.' . $field;
+            return $this->queryBuilder->getRootAliases()[0] . '.' . $field;
```
```suggestion
            return $this->getRootAlias()()[0] . '.' . $field;
```

pamil commented:
Thanks, Grzegorz! :tada:
